xarray/coding/times.py

```diff
@@ -266,20 +265,75 @@ def _decode_datetime_with_pandas(
     time_units, ref_date_str = _unpack_netcdf_time_units(units)
     time_units = _netcdf_to_numpy_timeunit(time_units)
     try:
-        # TODO: the strict enforcement of nanosecond precision Timestamps can be
-        # relaxed when addressing GitHub issue #7493.
-        ref_date = nanosecond_precision_timestamp(ref_date_str)
+        # relaxed to non-nanosecond resolution
+        ref_date = pd.Timestamp(ref_date_str)
+        # strip tz information
+        if ref_date.tz is not None:
+            ref_date = ref_date.tz_convert(None)
+        # get default unit and delta
+        default_unit = _get_datetime_resolution()
+        default_delta = np.timedelta64(1, default_unit).astype("timedelta64[ns]")
+        # get ref_date and time delta
+        ref_date_unit = (
+            ref_date.unit
+            if hasattr(ref_date, "unit")
+            else np.datetime_data(ref_date.asm8)[0]
+        )
+        ref_date_delta = np.timedelta64(1, ref_date_unit).astype("timedelta64[ns]")
+        time_delta = np.timedelta64(1, time_units).astype("timedelta64[ns]")
```

Comment:
From my perspective the **decoding process** should just do what it is supposed to do , which is decode into the designated datetimes as given by `time_unit since reference_date`. As our final dtype (`np.datetime64`) is based on unix epoch (`1970-01-01`, with different possible resolutions) we need to find the fitting resolution.

`days since 1970-01-01 00:00:00` -> `"s"` (reference_date takes precedence over time_unit since it is in second resolution)
`seconds since 1970-01-01` ->  `"s"` (time_unit take precedence over reference_date)
`milliseconds since 1970-01-01 00:00:00.000000001` ->  `"ns"` (reference_date takes precedence over time_unit)

For now we have a third possible element to consider, the default resolution, which currently is "ns" (or might be changed to something else now). 

The current proposed workflow with the `default resolution` should finally not be needed any more, when all issues have been ironed out. Or, we use something like `None` to specify automatic inferring.

> But maybe another question at this stage is whether we would want the encoding information (reference date or time units) to affect the datetime precision we decode to—currently it appears one could end up with non-default precision if the time units in the file specified something finer than the user-specified precision via `set_options`, which could be surprising.

Maybe the idea of a default resolution should only come into consideration in the encoding process, when the user does not specify any `units`?

Reply:
Or no default resolution at all and try to infer everything (if that's possible at all)?

A good example showing the ambiguities between on-disk and in-memory is something like

- on-disk:   `[0, 1, 2, 3]` with `days since 1970-01-01 12:00:00` 
will translate to:
- in-memory: `[12, 36, 60, 84]` with `datetime64[h]`
- in_memory default `"s"`:  `[ 43200, 129600, 216000, 302400]` with `datetime64[s]`

A default resolution of seconds `s` would be not problem for the above, but for this example:

- on-disk:   [0, 1, 2, 3] with `seconds since 1970-01-01 00:00:00.001` 
will translate to:
- in-memory: [1, 1001, 2001, 3001] with `datetime64[ms]`
- in_memory default `"s"`: `[0, 1, 2, 3]` with `datetime64[s]` 

we would run into issues. The more I think about this, the more think inferring as much as possible is the way to go, to not round/cutoff or otherwise get time mismatches between on-disk and in-memory.

Reply:
Indeed I agree long-term some kind of inference would be best—ideally we do not lose precision based on what was written on disk.  It is just somewhat complicated (in the case of floating point times, one would need to take into account the values themselves in addition to the metadata), but we can try to hash out what that could look like (or punt and just say floats are always decoded to nanosecond precision).

For context, I interpreted this PR as temporarily aiming to maintain a consistent NumPy datetime precision throughout xarray, while at least allowing for some flexibility (hence the modifications to the casting logic in `variable.py` to also respect the `time_resolution` set in `set_options`), so it was surprising to see one place where datetime precisions other than that specified in `set_options` could appear.  

In your eyes, at what point do we remove the casting logic and fully allow mixed precision datetime values?  Is this merely just an issue with the tests at this stage?

Reply:
@spencerkclark Sorry, went down that rabbit hole the last days. I think I'm on a good way to have things sorted out. There are different locations in the code and tests were the gears mesh together. 

The overall decoding and encoding logic seem to work in my local feature branch. I'll update this PR when I've fixed all the bits and pieces. Need to dig further now, back into the hole :rabbit:

Reply:
No worries—if at any point you feel it would be helpful to discuss anything in a more interactive way we could try to find a time to meet.  Happy to keep things asynchronous / open though too.  Thanks again for your work on this!